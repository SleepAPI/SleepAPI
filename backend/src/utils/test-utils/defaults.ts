import { CustomPokemonCombinationWithProduce } from '@src/domain/combination/custom';
import { Produce } from '@src/domain/combination/produce';
import { ProductionStats } from '@src/domain/computed/production';
import { TimePeriod } from '@src/domain/time/time';
import { berry, ingredient, mainskill, nature, pokemon, subskill } from 'sleepapi-common';
import { getEmptyProduce } from '../production-utils/production-utils';

export const MOCKED_MAIN_SLEEP: TimePeriod = {
  start: {
    hour: 6,
    minute: 0,
    second: 0,
  },
  end: {
    hour: 21,
    minute: 30,
    second: 0,
  },
};

export const MOCKED_OPTIMAL_PRODUCTION_STATS: ProductionStats = {
  level: 60,
  nature: nature.QUIET,
  subskills: [subskill.HELPING_SPEED_M, subskill.INGREDIENT_FINDER_M, subskill.INGREDIENT_FINDER_S],
  e4e: 0,
  cheer: 0,
  extraHelpful: 0,
  helperBoostUnique: 1,
  helperBoostProcs: 0,
  helpingBonus: 0,
  camp: false,
  erb: 0,
  incense: false,
  skillLevel: 6,
  mainBedtime: MOCKED_MAIN_SLEEP.end,
  mainWakeup: MOCKED_MAIN_SLEEP.start,
};

export const MOCKED_PRODUCE: Produce = {
  berries: {
    amount: 2,
    berry: berry.GREPA,
  },
  ingredients: [
    {
      amount: 1,
      ingredient: ingredient.FANCY_APPLE,
    },
  ],
};

export const MOCKED_POKEMON: pokemon.Pokemon = {
  berry: berry.BELUE,
  carrySize: 20,
  frequency: 2500,
  ingredient0: {
    amount: 1,
    ingredient: ingredient.BEAN_SAUSAGE,
  },
  ingredient30: [
    {
      amount: 2,
      ingredient: ingredient.FANCY_APPLE,
    },
  ],
  ingredient60: [
    {
      amount: 3,
      ingredient: ingredient.FANCY_EGG,
    },
  ],
  ingredientPercentage: 20,
  maxCarrySize: 20,
  name: 'MOCK_POKEMON',
  skill: mainskill.CHARGE_STRENGTH_M,
  skillPercentage: 2,
  specialty: 'skill',
};

export const MOCKED_POKEMON_WITH_PRODUCE: CustomPokemonCombinationWithProduce = {
  pokemonCombination: {
    pokemon: pokemon.PINSIR,
    ingredientList: [
      { amount: 2, ingredient: ingredient.HONEY },
      { amount: 5, ingredient: ingredient.FANCY_APPLE },
    ],
  },
  detailedProduce: {
    produce: {
      berries: {
        amount: 0,
        berry: berry.LUM,
      },
      ingredients: [
        { amount: 2, ingredient: ingredient.HONEY },
        { amount: 5, ingredient: ingredient.FANCY_APPLE },
      ],
    },
    spilledIngredients: [],
    sneakySnack: {
      amount: 0,
      berry: berry.LUM,
    },
    dayHelps: 0,
    nightHelps: 0,
    averageTotalSkillProcs: 0,
    skillActivations: [],
  },
  averageProduce: getEmptyProduce(berry.LUM),
  customStats: { level: 30, nature: nature.RASH, subskills: [], skillLevel: 6 },
};
