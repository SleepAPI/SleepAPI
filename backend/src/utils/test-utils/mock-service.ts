/* eslint-disable @typescript-eslint/no-explicit-any */
type RecordedCall = {
  name: string;
  method: string | number | symbol;
  arguments: Array<unknown>;
};

type RecordedCalls = Array<RecordedCall>;
type RecordedCallPather = (call: RecordedCall) => RecordedCall;

class MockServiceImpl {
  #savedStatus: Array<{ target: object; state: object }> = [];
  #recordedCalls: RecordedCalls = [];
  #recordedCallPatchers = new Array<RecordedCallPather>();

  public init(objects: Record<string, object>) {
    Object.values(objects).forEach((target) => {
      this.#savedStatus.push({ target, state: { ...target } });
    });
  }

  public restore() {
    this.#savedStatus.forEach((saved) => {
      Object.keys(saved.target).forEach((key) => delete (saved.target as any)[key]);
      Object.assign(saved.target, saved.state);
    });
    this.#savedStatus = [];
    this.#recordedCalls = [];
    this.#recordedCallPatchers = [];
  }

  public installRecordPather(patcher: RecordedCallPather) {
    this.#recordedCallPatchers.push(patcher);
  }

  public get recorded() {
    const patchCalls = (recordedCall: RecordedCall) => {
      let result = recordedCall;
      for (const recordedCallPatcher of this.#recordedCallPatchers) {
        result = recordedCallPatcher(result);
      }
      return result;
    };

    return this.#recordedCalls.map(patchCalls).map((recordedCall) => ({
      [`${recordedCall.name}.${String(recordedCall.method)}`]: recordedCall.arguments,
    }));
  }

  public resetRecordings() {
    this.#recordedCalls = [];
  }

  public record<MOCK_TARGET extends object>(targets: Record<string, MOCK_TARGET>): MOCK_TARGET {
    if (Object.values(targets).length !== 1) {
      throw new Error('Only one object may be mocked per invocation to mock');
    }
    const [name, target] = Object.entries(targets)[0];
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const self = this;
    return new Proxy(target, {
      set(t: any, p, newValue) {
        if (typeof newValue === 'function') {
          t[p] = (...args: any[]) => {
            self.#recordedCalls.push({
              name,
              method: p,
              arguments: args,
            });
            return newValue(...args);
          };
        }
        return true;
      },
    });
  }
}

export const MockService = new MockServiceImpl();
