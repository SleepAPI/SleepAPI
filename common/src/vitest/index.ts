export * as commonMocks from './mocks';
