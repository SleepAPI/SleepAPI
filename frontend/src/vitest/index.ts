export * from './mocks'
export * from './setup'
