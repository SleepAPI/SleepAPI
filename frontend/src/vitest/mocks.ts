import { ingredient, nature, pokemon, type PokemonInstanceExt } from 'sleepapi-common'

export function createMockPokemon(attrs?: Partial<PokemonInstanceExt>): PokemonInstanceExt {
  return {
    name: 'Bubbles',
    externalId: 'external-id',
    pokemon: pokemon.PIKACHU,
    carrySize: 10,
    ingredients: [{ level: 1, ingredient: ingredient.FANCY_APPLE }],
    level: 10,
    ribbon: 0,
    nature: nature.BASHFUL,
    saved: false,
    shiny: false,
    skillLevel: 1,
    subskills: [],
    version: 1,
    ...attrs
  }
}
