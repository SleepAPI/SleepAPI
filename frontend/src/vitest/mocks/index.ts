export * from './calculator'
export * from './pokemon-instance'
