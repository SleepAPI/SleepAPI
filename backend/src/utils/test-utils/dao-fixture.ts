import knex, { Knex } from 'knex';
import { DatabaseService } from '../../database/database-service';
import DatabaseMigration from '../../database/migration/database-migration';
import { MockService } from './mock-service';

type InitParams = {
  enforceForeignKeyConstraints?: boolean;
  recreateDatabasesBeforeEachTest?: boolean;
};

async function truncateAllTables(params?: InitParams, knex?: Knex): Promise<void> {
  if (!knex) {
    return;
  }

  const SQLITE_SEQUENCE_TABLE = 'sqlite_sequence';
  const EXCLUDE_PREFIX = ['knex', 'sqlite'];

  const tables = await knex('sqlite_master').where('type', 'table');
  const tableNames = tables
    ?.map((row) => row.name)
    .filter((tableName) => !EXCLUDE_PREFIX.some((prefix) => tableName.startsWith(prefix)));

  await knex.raw('PRAGMA foreign_keys = OFF');

  await Promise.all(tableNames?.map((tableName) => knex(tableName).truncate()) ?? []);
  if (await knex.schema.hasTable(SQLITE_SEQUENCE_TABLE)) {
    await knex(SQLITE_SEQUENCE_TABLE).truncate();
  }

  if (params?.enforceForeignKeyConstraints) {
    await knex?.raw('PRAGMA foreign_keys = ON');
  }
}

export const DaoFixture = {
  init(params?: InitParams) {
    let pokemonsleepDB: Knex;

    if (params?.recreateDatabasesBeforeEachTest) {
      beforeEach(async () => {
        await setup();
      });
    } else {
      beforeAll(async () => {
        await setup();
      });
    }

    async function setup() {
      MockService.init({ DatabaseService });

      pokemonsleepDB = knex({
        client: 'sqlite3',
        useNullAsDefault: true,
        connection: ':memory:',
        acquireConnectionTimeout: 10000,
      });

      Object.defineProperty(DatabaseService, '#knex', {
        configurable: true,
        value: pokemonsleepDB,
      });

      if (params?.enforceForeignKeyConstraints) {
        await pokemonsleepDB?.raw('PRAGMA foreign_keys = ON');
      }

      DatabaseService.getKnex = () => Promise.resolve(pokemonsleepDB);
      await DatabaseMigration.migrate();
    }

    async function destroyDatabases() {
      await pokemonsleepDB.destroy();
    }

    afterEach(async () => {
      if (params?.recreateDatabasesBeforeEachTest) {
        await destroyDatabases();
      } else {
        await truncateAllTables(params, pokemonsleepDB);
      }

      MockService.restore();
    });

    afterAll(async () => {
      if (!params?.recreateDatabasesBeforeEachTest) {
        await destroyDatabases();
      }
    });
  },
};
