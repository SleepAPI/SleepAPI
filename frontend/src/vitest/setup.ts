import router from '@/router/router'
import { config } from '@vue/test-utils'
import ResizeObserver from 'resize-observer-polyfill'
import 'vitest-canvas-mock'
import { createVuetify } from 'vuetify'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'
import { VTimePicker } from 'vuetify/labs/VTimePicker'
import 'vuetify/styles'

const vuetify = createVuetify({
  components: { ...components, VTimePicker },
  directives
})

config.global.plugins.push(vuetify, router)

config.global.stubs = {
  RouterView: true,
  RouterLink: true
}

global.ResizeObserver = ResizeObserver
