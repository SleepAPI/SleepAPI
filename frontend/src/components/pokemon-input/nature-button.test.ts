import NatureButton from '@/components/pokemon-input/nature-button.vue'
import { VueWrapper, mount } from '@vue/test-utils'
import { nature } from 'sleepapi-common'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'

describe('NatureButton', () => {
  let wrapper: VueWrapper<InstanceType<typeof NatureButton>>
  const sampleNature = nature.BASHFUL

  beforeEach(() => {
    wrapper = mount(NatureButton, {
      props: {
        nature: sampleNature
      }
    })
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('renders correctly with provided data', () => {
    expect(wrapper.find('.responsive-text').text()).toContain(sampleNature.name)
  })

  it('computes natureName correctly when enabled', () => {
    expect(wrapper.vm.natureName).toBe(sampleNature.name)
  })

  it('updates nature and emits update-nature event when selectNature is called', () => {
    const newNatureName = nature.BRAVE
    wrapper.vm.selectNature(newNatureName)
    const emitted = wrapper.emitted('update-nature') as Array<Array<nature.Nature>>
    expect(emitted).toHaveLength(1)

    const emittedEvent = emitted[0][0]
    expect(emittedEvent.name).toBe(newNatureName.name)
  })

  it('opens the dialog when the button is clicked', async () => {
    const button = wrapper.find('button')
    await button.trigger('click')
    expect(wrapper.vm.natureMenu).toBe(true)
    expect(wrapper.findComponent({ name: 'NatureMenu' }).exists()).toBe(true)
  })
})
