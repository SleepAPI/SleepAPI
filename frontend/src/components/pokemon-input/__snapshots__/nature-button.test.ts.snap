// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NatureButton > computes filteredNatures correctly 1`] = `
[
  {
    "category": "Speed",
    "list": [
      "Lonely",
      "Adamant",
      "Naughty",
      "Brave",
    ],
  },
  {
    "category": "Ingredient",
    "list": [
      "Modest",
      "Mild",
      "Rash",
      "Quiet",
    ],
  },
  {
    "category": "Skill",
    "list": [
      "Calm",
      "Gentle",
      "Careful",
      "Sassy",
    ],
  },
  {
    "category": "Energy",
    "list": [
      "Bold",
      "Impish",
      "Lax",
      "Relaxed",
    ],
  },
  {
    "category": "EXP",
    "list": [
      "Timid",
      "Hasty",
      "Jolly",
      "Naive",
    ],
  },
  {
    "category": "Neutral",
    "list": [
      "Bashful",
      "Hardy",
      "Docile",
      "Quirky",
      "Serious",
    ],
  },
]
`;
